client/state/data-layer/http-data/getters.js

```diff
@@ -0,0 +1,20 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import { http as rawHttp } from 'state/http/actions';
+import { requestHttpData } from 'state/data-layer/http-data';
+
+export const requestGeoLocation = () =>
+	requestHttpData(
+		'geo',
+		rawHttp( {
+			method: 'GET',
+			url: 'https://public-api.wordpress.com/geo/',
+		} ),
+		{
+			fromApi: () => ( { body: { country_short } } ) => [
+				[ 'geo', parseInt( country_short, 10 ) ],
```

Comment:
I see that `fromApi` has the power to update any key it wishes, not only `geo`, the requested one. Or even multiple keys. Is there a use case for this flexibility?

Reply:
the use case may be more strongly related to my original intention to support collections but also provides any related data we want in a mostly-flat structure.

for example, if we grab a post from the post endpoint we can update the post-like value as well.

I don't use it here yet but in every iteration where I tried I was unable to get very far without allowing arbitrary data updates. any time we grab a collection of things we need to set a different value…

```js
fromApi: () => ( { posts } ) => [
	...posts.map( post => [ `posts-${ post.id }`, post ] ),
	[ 'available-posts', posts.map( ( { id } ) => id )
];
```

collections are possible here without the `type/item` distinction in some of my more filled-out-but-more-complicated incantations by use of storing the items by id and providing a group dataset as in the example above.

it's worth noting here since the question will come up eventually - what about updates vs. replacements? I figured that we can provide a function option here which feeds in the old value but so far that's not yet necessary as we have `getHttpData()` which can do the same thing. until that becomes a more obvious use-case I'm leaving it out, scoping this down.

hope that is a helpful response!

Reply:
Thanks, that's definitely a helpful response! Requesting a collection item would look like this, I suppose?
```js
const requestPost = id => requestHttpData( `post-${id}`, ... );
```

Reply:
in my mind it would probably end up with using either a string for a single value or a pair of strings for a type and id

```js
const post = requestHttpData( [ `site-posts-${ siteId }`, postId ], … );
const site = requestHttpData( `site-${ siteId }`, … );
```

By not using one string it gives us the change to do a quick "grab me the collection"

```js
export const getHttpData = id => {
	if ( 'string' === typeof id ) {
		return httpData.get( id );
	}

	const [ type, item ] = id;

	return item
		? httpData.get( type ).get( item )
		: httpData.get( type ).values()
}
```

^^^ something like that snippet but actually filled out - the snippet is for the concept not the implementation

Reply:
note that this is all sugar over the current flat hierarchy which could possibly prove a better long-term decision because it constrains us to think in flat terms.

```js
const post = requestHttpData( `post-${ siteId }-${ postId }`, … );
const posts = requestHttpData( `site-posts-${ siteId }`, … )
	.data
	.map( postId => requestHttpData( `post-${ siteId }-${ postId }` ) );
```